### DISCUSSION TITLE
Requiring modules like Node.js results in an access denied error message

### DISCUSSION THREAD

**Discussion by dejannenov** [denoland/deno#9534] [Q&A]
Trying trivial snippet, which runs fine in node, but fails with deno:

typescript:

```ts
import axios from './node_modules/axios';
const url = 'https://jsonplaceholder.typicode.com/todos/1'

axios.get(url).then(response => {
    console.log(response.data);
});
```

javascript generated with tsc version 4.1.5:

```js
"use strict";
exports.__esModule = true;
var axios_1 = require("axios");
var url = 'https://jsonplaceholder.typicode.com/todos/1';
axios_1["default"].get(url).then(function (response) {
    console.log(response.data);
});
```

From Windows 10 Professional Power Shell launched as Adminsitrator:

With node:

```
PS D:\dev\ts> node .\getjson.js
{ userId: 1, id: 1, title: 'delectus aut autem', completed: false }
```

With deno:

```
PS D:\dev\ts> deno run .\getjson.ts -A 
Compile file:///D:/dev/ts/getjson.ts
error: Uncaught PermissionDenied: Access is denied. (os error 5)
    at unwrapResponse ($deno$/ops/dispatch_json.ts:43:11)
    at Object.sendAsync ($deno$/ops/dispatch_json.ts:98:10)
    at async processImports ($deno$/compiler.ts:736:23)
    at async processImports ($deno$/compiler.ts:753:7)
    at async compile ($deno$/compiler.ts:1316:31)
    at async tsCompilerOnMessage ($deno$/compiler.ts:1548:22)
    at async workerMessageRecvCallback ($deno$/runtime_worker.ts:74:9)
```

**Comment by kitsonk** [comment]
You should read the Deno manual.  Specifically: https://deno.land/manual@v1.7.4/examples/import_export

Deno does no "magical" module resolution, unlike Node.js.  So instead of importing a module with `'./node_modules/axios'` you attempt to import a directory, which you cannot do.

Most Node.js package aren't going to work when installed locally.  Consider a CDN for sourcing these modules (like for example `https://www.skypack.dev/view/axios?dts` which would also include all the TypeScript types for you).  But even then most people don't use axios with Deno because Deno has a browser compatible `fetch()` built into it.

**Reply by dejannenov** [reply]
My bad - I do need to RTFM :(

Please be aware, however, that we are "selling" the value of deno as an "improved/fixed node.js"

If we have to essentially re-factor / re-write the thousands of man-years of node code ... that will be much more challenging.

In my particular SNAFU here - it seems it would be easy to identify an unsupported node module import and produce that as an error message, rather than a cryptic "os error 5"? 

To make a 1/2 as--d excuse: I have come to rely on rustc to help me with trivial (and not so trivial) mistakes and had mentally / by association extended the benefit of the doubt to deno as well - because it compiled and ran my node-module-polluted code, I expected it to run.  Clearly, assumption is still the mother of all f-ups :)

May I also suggest that on the home page of deno.land, right under:

"Has a set of reviewed (audited) standard modules that are guaranteed to work with Deno: deno.land/std"

the following (or some such) text be added:

"No npm module support. Do not attempt to use node modules. This prevents dependency messes and security issues, etc."

**Comment by ghost** [comment]
That said, if you really want to require modules like node.js, you can just write your own require and use an `import-map.json` for filename resolution:

```js
async function require(path){
return await import(path)
}
```

**Reply by ghost** [reply]
This would even allow you to rebuild node on top of deno. Pretty cool...

**Reply by dejannenov** [reply]
I like it!

**Reply by Soremwar** [reply]
A require polyfill is already implemented on std

Check https://deno.land/std@0.88.0/node and https://deno.land/std@0.88.0/node/module.ts

**Reply by AddictArts** [reply]
> A require polyfill is already implemented on std
> 
> Check https://deno.land/std@0.88.0/node and https://deno.land/std@0.88.0/node/module.ts

How do you even use the poly fill? There is nothing on the page to help. Thanks

**Reply by bartlomieju** [reply]
@AddictArts you can now import built-in Node.js modules directly in Deno, like so:
```
import {createRequire} from "node:module";
const require = createRequire(import.meta.url);
require(...);
```

**Reply by AddictArts** [reply]
> @AddictArts you can now import built-in Node.js modules directly in Deno, like so:
> 
> ```
> import {createRequire} from "node:module";
> const require = createRequire(import.meta.url);
> require(...);
> ```

Thank you. I ended up importing index.ts and adding the tsconfig compiler option `"allowImportingTsExtensions": true`. I also leveraged `import_map.json` and `deno.jsonc` to reference it.

**Comment by Jeremie-Kassianoff** [comment]
Axios recently had a serious security issue, so it’s generally safer to avoid adding it unless you absolutely need it. 

Deno provides a built‑in, browser‑compatible fetch() API, you don’t need any external dependency for HTTP requests.

Using fetch reduces your attack surface and avoids relying on packages that may introduce additional vulnerabilities.
